core/src/main/java/com/graphhopper/routing/ProfileResolver.java

```diff
@@ -0,0 +1,195 @@
+                    match1.getWeighting().getFlagEncoder().toString().equals(match2.getWeighting().getFlagEncoder().toString()) &&
+                    match1.isEdgeBased() != match2.isEdgeBased()) {
+                return match1.isEdgeBased() ? match1 : match2;
+            }
+            throw new IllegalArgumentException("There are multiple CH profiles matching your request. Use the `weighting`,`vehicle`,`edge_based` and/or `u_turn_costs` parameters to be more specific." +
+                    "\nYou can also try disabling CH altogether using " + Parameters.CH.DISABLE + "=true" +
+                    "\nrequested:  " + getCHRequestAsString(hintsMap, edgeBased, uTurnCosts) + "\nmatched:   " + matchingProfiles + "\navailable: " + chProfiles);
+        }
+    }
+
+    public LMProfile selectLMProfile(List<LMProfile> lmProfiles, HintsMap hintsMap) {
+        List<LMProfile> matchingProfiles = new ArrayList<>();
+        for (LMProfile p : lmProfiles) {
+            if (!lmProfileMatchesHints(p, hintsMap))
+                continue;
+            matchingProfiles.add(p);
+        }
+        // Note:
+        // There are situations where we can use the requested encoder/weighting with an existing LM preparation, even
+        // though the preparation was done with a different weighting. For example this works when the new weighting
+        // only yields higher (but never lower) weights than the one that was used for the preparation. However, its not
+        // trivial to check whether or not this is the case so we do not allow this for now.
+        if (matchingProfiles.isEmpty()) {
+            throw new IllegalArgumentException("Cannot find matching LM profile for your request. Please check your parameters." +
+                    "\nYou can try disabling LM by setting " + Parameters.Landmark.DISABLE + "=true" +
+                    "\nrequested: " + getLMRequestAsString(hintsMap) + "\navailable: " + lmProfilesAsStrings(lmProfiles));
+        } else if (matchingProfiles.size() == 1) {
+            return matchingProfiles.get(0);
+        } else {
+            throw new IllegalArgumentException("There are multiple LM profiles matching your request. Use the `weighting` and `vehicle` parameters to be more specific." +
+                    "\nYou can also try disabling LM altogether using " + Parameters.CH.DISABLE + "=true" +
+                    "\nrequested:  " + getLMRequestAsString(hintsMap) + "\nmatched:   " + lmProfilesAsStrings(matchingProfiles) + "\navailable: " + lmProfilesAsStrings(lmProfiles));
+        }
+    }
+
+    protected boolean chProfileMatchesHints(CHProfile p, HintsMap hintsMap) {
+        Boolean edgeBased = getEdgeBased(hintsMap);
+        Integer uTurnCosts = getUTurnCosts(hintsMap);
+        return (edgeBased == null || p.isEdgeBased() == edgeBased) &&
+                (uTurnCosts == null || p.getWeighting().getTurnCostProvider().getName().equals(uTurnCosts.toString())) &&
```

Comment:
I do not like how we rely on the turn cost provider name to contain the u-turn costs here, but so far I do not have a better idea to do this. Also this is only needed to cover a (probably exotic) case: Multiple edge-based profiles with different u-turn costs. Once we have the profile request parameter we 'only' need to support this u-turn costs comparison here for legacy reasons, so this makes me even more want to get rid of it.

Reply:
Yes, probably we have to accept that ProfileResolver is mainly for backward compatibility? The bigger problem is the new getTurnCostProvider method, which I would prefer to be private.

Reply:
Yes, I would like to remove/not add `getTurnCostProvider` as well. But how else can we check what u-turn costs a weighting is using? Maybe we can skip this selection rule and throw an error if there are multiple edge-based weightings for the given weighting and vehicle. In this case one would *have* to use the profile parameter when using multiple edge-based CH preparations with different u-turn costs.

Reply:
> Yes, probably we have to accept that ProfileResolver is mainly for backward compatibility?

Yes absolutely, maybe not yet in this PR but `ProfileResolver` *only* exists for backward-compatibility.

Reply:
> In this case one would have to use the profile parameter when using multiple edge-based CH preparations with different u-turn costs.

This sounds not too bad.

Reply:
Done